src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -6355,6 +6408,416 @@ register_handler(int sig, struct sigaction *old_handler)
 	}
 }
 
+/* free all EQs allocated for aio */
+
+static void
+free_aio_ctx(void)
+{
+	int i;
+
+	for (i = 0; i < aio_eq_count; i++) {
+		daos_eq_destroy(aio_eq_list[i].eq, 0);
+		D_MUTEX_DESTROY(&aio_eq_list[i].lock);
+	}
+	aio_eq_count = 0;
+}
+
+/* aio functions return negative errno in case of failure */
+static int
+create_ev_eq_for_aio(aio_ctx_t *aio_ctx);
+int
+io_setup(int maxevents, io_context_t *ctxp)
+{
+	int        rc;
+	aio_ctx_t *aio_ctx_obj;
+
+	if (next_io_setup == NULL) {
+		next_io_setup = dlsym(RTLD_NEXT, "io_setup");
+		D_ASSERT(next_io_setup != NULL);
+	}
+
+	rc = next_io_setup(maxevents, ctxp);
+	if (rc < 0)
+		return rc;
+
+	D_ALLOC(aio_ctx_obj, sizeof(aio_ctx_t));
+	if (aio_ctx_obj == NULL) {
+		io_destroy(*ctxp);
+		return -ENOMEM;
+	}
+	aio_ctx_obj->ctx   = *ctxp;
+	aio_ctx_obj->depth = maxevents;
+	atomic_store_relaxed(&aio_ctx_obj->num_op_submitted, 0);
+	atomic_store_relaxed(&aio_ctx_obj->num_op_done, 0);
+	aio_ctx_obj->inited       = false;
+	aio_ctx_obj->on_dfs       = false;
+	aio_ctx_obj->ev_done_list = NULL;
+	aio_ctx_obj->ev_done_h    = 0;
+	aio_ctx_obj->ev_done_t    = 0;
+	aio_ctx_obj->ev_queue_len = 0;
+	*ctxp                     = (io_context_t)aio_ctx_obj;
+
+	rc = D_MUTEX_INIT(&aio_ctx_obj->lock, NULL);
+	if (rc) {
+		io_destroy(*ctxp);
+		D_FREE(aio_ctx_obj);
+		return -daos_der2errno(rc);
+	}
+
+	/* assume all io requests are over DFS for now. */
+	rc = create_ev_eq_for_aio(aio_ctx_obj);
+	if (rc) {
+		return (-rc);
+	}
+
+	return 0;
+}
+
+int
+io_destroy(io_context_t ctx)
+{
+	aio_ctx_t   *aio_ctx_obj = (aio_ctx_t *)ctx;
+	io_context_t ctx_save    = aio_ctx_obj->ctx;
+
+	if (next_io_destroy == NULL) {
+		next_io_destroy = dlsym(RTLD_NEXT, "io_destroy");
+		D_ASSERT(next_io_destroy != NULL);
+	}
+	D_MUTEX_DESTROY(&aio_ctx_obj->lock);
+	D_FREE(aio_ctx_obj->ev_done_list);
+	D_FREE(aio_ctx_obj);
+	return next_io_destroy(ctx_save);
+}
+
+static int
+create_ev_eq_for_aio(aio_ctx_t *aio_ctx)
+{
+	int rc, i, j, idx;
+	int num_aio_eq_free, num_aio_eq_create;
+
+	D_MUTEX_LOCK(&lock_aio);
```

Comment:
you are not locking the aio context here, but this is a global lock to manage EQs for aio context, so please rename that to aio_eqs_lock_g

Reply:
Thank you! "lock_aio" now is renamed as "lock_aio_eq_g".

Reply:
ok. I will update in next commit. Thank you!